File: Runtime/php/libs/Phrost/IPCClient.php
<?php

namespace Phrost;
use Exception;

/**
 * Phrost IPC Client
 *
 * This class handles all the low-level, cross-platform socket/pipe
 * communication with the Swift IPC server.
 */
class IPCClient
{
    /** @var resource|Socket|null */
    private $pipe = null;
    private bool $isWindows;
    private bool $isConnected = false;

    public function __construct()
    {
        $this->isWindows = strtoupper(substr(PHP_OS, 0, 3)) === "WIN";
    }

    /**
     * Connects to the Swift IPC server.
     * @throws Exception if connection fails.
     */
    public function connect(): void
    {
        if ($this->isConnected) {
            return;
        }

        if ($this->isWindows) {
            // --- Windows Connection (Named Pipe) ---
            $pipePath = "\\\\.\\pipe\\PhrostEngine";
            echo "Attempting to connect to Windows pipe: $pipePath...\n";
            $this->pipe = @fopen($pipePath, "r+");

            if (!$this->pipe) {
                throw new Exception(
                    "Failed to open pipe. Is the Swift server (PhrostIPC.exe) running?\n",
                );
            }
            stream_set_blocking($this->pipe, true);
        } else {
            // --- macOS/Linux Connection (UNIX Domain Socket) ---
            $pipePath = "/tmp/PhrostEngine.socket";
            echo "Attempting to connect to UNIX socket: $pipePath...\n";

            if (!extension_loaded("sockets")) {
                throw new Exception(
                    "The 'sockets' extension is not enabled in your php.ini.",
                );
            }

            $this->pipe = @socket_create(AF_UNIX, SOCK_STREAM, 0);
            if (!$this->pipe) {
                throw new Exception(
                    "socket_create() failed: " .
                    socket_strerror(socket_last_error()),
                );
            }

            $connected = @socket_connect($this->pipe, $pipePath);
            if (!$connected) {
                throw new Exception(
                    "socket_connect() failed: " .
                    socket_strerror(socket_last_error($this->pipe)) .
                    ". Is the Swift server running?\n",
                );
            }
            socket_set_block($this->pipe);
        }

        $this->isConnected = true;
        echo "Connected! Entering game loop...\n";
        echo "Blocking now...\n";
    }

    /**
     * Disconnects from the server.
     */
    public function disconnect(): void
    {
        if (!$this->isConnected || !$this->pipe) {
            return;
        }

        if ($this->isWindows) {
            if (is_resource($this->pipe)) {
                fclose($this->pipe);
            }
        } else {
            if (is_resource($this->pipe) || $this->pipe instanceof \Socket) {
                socket_close($this->pipe);
            }
        }
        $this->pipe = null;
        $this->isConnected = false;
        echo "Disconnected.\n";
    }

    /**
     * Runs the main game loop, calling the update callback each frame.
     *
     * @param callable $updateCallback The user's game logic function, e.g., 'Phrost_Update'
     * It must accept: (int $elapsed, float $dt, string $eventsBlob)
     * It must return: (string $commandBlob)
     */
    public function run(callable $updateCallback): void
    {
        if (!$this->isConnected) {
            throw new Exception("Cannot run: Not connected.");
        }

        $elapsed = 0;

        try {
            while (true) {
                // 1. Read frame data from Swift
                $frameData = $this->readFrame();
                if ($frameData === false) {
                    echo "Pipe broken (read failed). Exiting loop.\n";
                    break;
                }

                // 2. Call the user's game logic function
                $commandBlob = $updateCallback(
                    $elapsed,
                    $frameData["dt"],
                    $frameData["eventsBlob"],
                );

                if ($commandBlob === false) {
                    echo "[PHP Logic] Game logic signaled graceful quit.\n";
                    break;
                }

                // 3. Write commands back to Swift
                if ($this->writeFrame($commandBlob) === false) {
                    echo "Pipe broken (write failed). Exiting loop.\n";
                    break;
                }

                $elapsed++;
            }
        } catch (Exception $e) {
            echo "An error occurred during the loop: " .
                $e->getMessage() .
                "\n";
            error_log(
                "PHP Client Exception: " .
                $e->getMessage() .
                "\n" .
                $e->getTraceAsString(),
            );
        }
    }

    /**
     * Reads one full "frame" of data from the Swift server.
     * @return array|false ['dt' => float, 'eventsBlob' => string] or false on failure
     */
    private function readFrame(): bool|array
    {
        // 1. Read Length Header
        $lenHeader = $this->read_all(4);
        if ($lenHeader === false) {
            return false;
        }

        $unpackedHeader = unpack("VtotalLength", $lenHeader);
        if (!$unpackedHeader) {
            error_log("Failed to unpack header.");
            return false;
        }
        $totalLength = $unpackedHeader["totalLength"];

        // 2. Read Payload (dt + events)
        if ($totalLength < 8) {
            error_log("Payload too small: $totalLength bytes, expected >= 8.");
            return false;
        }

        // Read DT (8 bytes)
        $dtData = $this->read_all(8);
        if ($dtData === false) {
            error_log("Failed to read delta-time.");
            return false;
        }
        $dt = unpack("d", $dtData)[1];

        // Read Events (Remaining bytes)
        $eventPayloadLength = $totalLength - 8;
        $eventsBlob = "";
        if ($eventPayloadLength > 0) {
            $eventsBlob = $this->read_all($eventPayloadLength);
            if ($eventsBlob === false) {
                error_log("Failed to read event payload.");
                return false;
            }
        }

        return ["dt" => $dt, "eventsBlob" => $eventsBlob];
    }

    /**
     * Writes one full "frame" of commands to the Swift server.
     * @return bool true on success, false on failure
     */
    private function writeFrame(string $commandBlob): bool
    {
        $cmdLen = strlen($commandBlob);
        $outData = pack("V", $cmdLen) . $commandBlob;

        if ($this->write_all($outData) === false) {
            return false;
        }

        if ($this->isWindows) {
            fflush($this->pipe);
        }
        return true;
    }

    /**
     * Unified read function. Reads exactly $length bytes.
     * @return string|false
     */
    private function read_all($length): string|bool
    {
        if ($length == 0) {
            return "";
        }
        $buffer = "";
        $bytesRemaining = $length;
        while ($bytesRemaining > 0) {
            $data = "";
            if ($this->isWindows) {
                $data = @fread($this->pipe, $bytesRemaining);
            } else {
                $data = @socket_read(
                    $this->pipe,
                    $bytesRemaining,
                    PHP_BINARY_READ,
                );
            }

            if ($data === false) {
                $errorMsg = $this->isWindows
                    ? "fread failed"
                    : "socket_read failed: " .
                    socket_strerror(socket_last_error($this->pipe));
                error_log($errorMsg);
                return false;
            }
            if ($data === "") {
                error_log(
                    "Pipe closed (read 0 bytes). Bytes remaining: $bytesRemaining",
                );
                return false;
            }

            $buffer .= $data;
            $bytesRemaining -= strlen($data);
        }
        return $buffer;
    }

    /**
     * Unified write function. Writes all data.
     * @param string $data Data to write
     * @return int|false bytes written or false on failure
     */
    private function write_all($data): int|bool
    {
        $totalToWrite = strlen($data);
        $totalWritten = 0;

        while ($totalWritten < $totalToWrite) {
            $bytesToWrite = substr($data, $totalWritten);
            $written = 0;

            if ($this->isWindows) {
                $written = @fwrite($this->pipe, $bytesToWrite);
            } else {
                $written = @socket_write(
                    $this->pipe,
                    $bytesToWrite,
                    strlen($bytesToWrite),
                );
            }

            if ($written === false) {
                $errorMsg = $this->isWindows
                    ? "fwrite failed"
                    : "socket_write failed: " .
                    socket_strerror(socket_last_error($this->pipe));
                error_log($errorMsg);
                return false;
            }
            if ($written === 0) {
                error_log("Write returned 0 bytes.");
                return false;
            }
            $totalWritten += $written;
        }
        return $totalWritten;
    }
}

File: Runtime/python/Phrost/ipc_client.py
import os
import socket
import struct
import sys
from typing import Callable, Optional, Dict, Union


class IPCClient:
    """
    Phrost IPC Client (Python)

    This class handles all the low-level, cross-platform socket/pipe
    communication with the Swift IPC server.
    """

    def __init__(self):
        self.is_windows: bool = os.name == "nt"
        self.pipe: Optional[Union[socket.socket, "file"]] = None
        self.is_connected: bool = False

    def connect(self):
        """Connects to the Swift IPC server."""
        if self.is_connected:
            return

        try:
            if self.is_windows:
                # --- Windows Connection (Named Pipe) ---
                pipe_path = r"\\.\pipe\PhrostEngine"
                print(f"Attempting to connect to Windows pipe: {pipe_path}...")

                # 'r+b' is critical for binary read/write
                self.pipe = open(pipe_path, "r+b")
                # Windows pipes are blocking by default, similar to stream_set_blocking

            else:
                # --- macOS/Linux Connection (UNIX Domain Socket) ---
                pipe_path = "/tmp/PhrostEngine.socket"
                print(f"Attempting to connect to UNIX socket: {pipe_path}...")

                self.pipe = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                self.pipe.connect(pipe_path)
                # Sockets are blocking by default, similar to socket_set_block

            self.is_connected = True
            print("Connected! Entering game loop...")
            print("Blocking now...")

        except (FileNotFoundError, ConnectionRefusedError) as e:
            server_name = "PhrostIPC.exe" if self.is_windows else "Swift server"
            raise Exception(
                f"Connection failed. Is the {server_name} running?\nError: {e}"
            )
        except Exception as e:
            raise Exception(f"An unexpected error occurred: {e}")

    def disconnect(self):
        """Disconnects from the server."""
        if not self.is_connected or not self.pipe:
            return

        try:
            self.pipe.close()
        except Exception as e:
            print(f"Error during disconnect: {e}", file=sys.stderr)
        finally:
            self.pipe = None
            self.is_connected = False
            print("Disconnected.")

    def run(self, update_callback: Callable[[int, float, bytes], Union[bytes, bool]]):
        """
        Runs the main game loop, calling the update callback each frame.

        :param update_callback: The user's game logic function.
            Accepts: (elapsed: int, dt: float, events_blob: bytes)
            Returns: (command_blob: bytes) or False to quit.
        """
        if not self.is_connected:
            raise Exception("Cannot run: Not connected.")

        elapsed = 0
        try:
            while True:
                # 1. Read frame data from Swift
                frame_data = self.read_frame()
                if frame_data is None:
                    print("Pipe broken (read failed). Exiting loop.")
                    break

                # 2. Call the user's game logic function
                command_blob = update_callback(
                    elapsed, frame_data["dt"], frame_data["events_blob"]
                )

                if command_blob is False:
                    print("[Python Logic] Game logic signaled graceful quit.")
                    break

                # 3. Write commands back to Swift
                if not self.write_frame(command_blob):
                    print("Pipe broken (write failed). Exiting loop.")
                    break

                elapsed += 1
        except Exception as e:
            print(f"An error occurred during the loop: {e}", file=sys.stderr)
            import traceback

            traceback.print_exc()

    def read_frame(self) -> Optional[Dict[str, Union[float, bytes]]]:
        """
        Reads one full "frame" of data from the Swift server.
        Format: [4-byte length][8-byte double dt][event_blob]
        """
        # 1. Read Length Header (4 bytes, unsigned long, little-endian)
        len_header = self.read_all(4)
        if len_header is None:
            return None

        try:
            # '<L' = unsigned long, little-endian (matches PHP 'V')
            total_length = struct.unpack("<L", len_header)[0]
        except struct.error:
            print("Failed to unpack header.", file=sys.stderr)
            return None

        # 2. Read Payload (dt + events)
        if total_length < 8:
            print(
                f"Payload too small: {total_length} bytes, expected >= 8.",
                file=sys.stderr,
            )
            return None

        # Read DT (8 bytes, double)
        dt_data = self.read_all(8)
        if dt_data is None:
            print("Failed to read delta-time.", file=sys.stderr)
            return None

        # 'd' = double (matches PHP 'd')
        dt = struct.unpack("d", dt_data)[0]

        # Read Events (Remaining bytes)
        event_payload_length = total_length - 8
        events_blob = b""
        if event_payload_length > 0:
            events_blob = self.read_all(event_payload_length)
            if events_blob is None:
                print("Failed to read event payload.", file=sys.stderr)
                return None

        return {"dt": dt, "events_blob": events_blob}

    def write_frame(self, command_blob: bytes) -> bool:
        """
        Writes one full "frame" of commands to the Swift server.
        Format: [4-byte length][command_blob]
        """
        try:
            cmd_len = len(command_blob)
            # '<L' = unsigned long, little-endian (matches PHP 'V')
            out_data = struct.pack("<L", cmd_len) + command_blob
            return self.write_all(out_data)
        except Exception as e:
            print(f"write_frame failed: {e}", file=sys.stderr)
            return False

    def read_all(self, length: int) -> Optional[bytes]:
        """Unified read function. Reads exactly 'length' bytes."""
        if length == 0:
            return b""

        buffer = bytearray()
        bytes_remaining = length
        try:
            while bytes_remaining > 0:
                if self.is_windows:
                    data = self.pipe.read(bytes_remaining)
                else:
                    data = self.pipe.recv(bytes_remaining)

                if not data:
                    print(
                        f"Pipe closed (read 0 bytes). Bytes remaining: {bytes_remaining}",
                        file=sys.stderr,
                    )
                    return None

                buffer.extend(data)
                bytes_remaining -= len(data)
            return bytes(buffer)
        except (IOError, socket.error) as e:
            print(f"read_all failed: {e}", file=sys.stderr)
            return None

    def write_all(self, data: bytes) -> bool:
        """Unified write function. Writes all data."""
        try:
            if self.is_windows:
                # For Windows pipes opened as files, we loop the write
                # just like your PHP code to ensure all data is sent.
                total_written = 0
                total_to_write = len(data)
                view = memoryview(data)  # Avoids string copies
                while total_written < total_to_write:
                    written = self.pipe.write(view[total_written:])
                    if written is None or written == 0:
                        print("Write returned 0 bytes.", file=sys.stderr)
                        return False
                    total_written += written
                self.pipe.flush()  # Essential for Windows pipes
            else:
                # Sockets have 'sendall' which handles the loop for us
                self.pipe.sendall(data)
            return True
        except (IOError, socket.error) as e:
            print(f"write_all failed: {e}", file=sys.stderr)
            return False
